src/main/java/co/cask/common/AvroToStructuredTransformer.java

```diff
@@ -0,0 +1,148 @@
+
+import co.cask.cdap.api.data.format.StructuredRecord;
+import co.cask.cdap.api.data.format.UnexpectedFormatException;
+import co.cask.cdap.api.data.schema.Schema;
+import co.cask.hydrator.common.RecordConverter;
+import com.google.common.collect.Maps;
+import org.apache.avro.generic.GenericRecord;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.time.LocalDate;
+import java.time.LocalTime;
+import java.time.ZoneId;
+import java.time.ZoneOffset;
+import java.time.ZonedDateTime;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+
+/**
+ * Create StructuredRecords from GenericRecords
+ */
+public class AvroToStructuredTransformer extends RecordConverter<GenericRecord, StructuredRecord> {
+  private final Map<Integer, Schema> schemaCache = Maps.newHashMap();
+
+  public StructuredRecord transform(GenericRecord genericRecord) throws IOException {
+    org.apache.avro.Schema genericRecordSchema = genericRecord.getSchema();
+    return transform(genericRecord, convertSchema(genericRecordSchema));
+  }
+
+  @Override
+  public StructuredRecord transform(GenericRecord genericRecord, Schema structuredSchema) throws IOException {
+    StructuredRecord.Builder builder = StructuredRecord.builder(structuredSchema);
+    for (Schema.Field field : structuredSchema.getFields()) {
+      String fieldName = field.getName();
+      Object value = convertField(genericRecord.get(fieldName), field.getSchema());
+      if (value instanceof LocalDate) {
+        builder.setDate(fieldName, (LocalDate) value);
+      } else if (value instanceof LocalTime) {
+        builder.setTime(fieldName, (LocalTime) value);
+      } else if (value instanceof ZonedDateTime) {
+        builder.setTimestamp(fieldName, (ZonedDateTime) value);
+      } else {
+        builder.set(fieldName, value);
+      }
+    }
+    return builder.build();
+  }
+
+  @Override
+  @Nullable
+  protected Object convertField(Object field, Schema fieldSchema) throws IOException {
+    if (field == null) {
+      return null;
+    }
+
+    Schema.Type fieldType = fieldSchema.getType();
+    Schema.LogicalType logicalType = fieldSchema.getLogicalType();
+
+    if (logicalType != null) {
+      switch (logicalType) {
+        case DATE:
+          // date will be in yyyy-mm-dd format
+          return LocalDate.parse(field.toString());
+        case TIME_MILLIS:
+        case TIME_MICROS:
+          // time will be in hh:mm:ss format
+          return LocalTime.parse(field.toString());
+        case TIMESTAMP_MILLIS:
+        case TIMESTAMP_MICROS:
+          return getZonedDateTime((long) field);
+        default:
+          throw new UnexpectedFormatException("field type " + fieldType + " is not supported.");
+      }
+    }
+
+    // This casting is added because bigquery up casts number fields. For example:
+    // int becomes long, float becomes double
+    switch (fieldType) {
+      case INT:
+        if (field instanceof Integer) {
```

Comment:
is this needed then? Won't it always be a long? If we're unsure about the actual type, it's better to cast to a Number and call intValue(), longValue(), etc where needed